src/backend/metal/src/window.rs

```diff
@@ -152,16 +164,23 @@ impl Drop for Swapchain {
 impl Swapchain {
     /// Returns the drawable for the specified swapchain image index,
     /// marks the index as free for future use.
-    pub(crate) fn take_drawable(&self, index: hal::SwapImageIndex) -> metal::Drawable {
+    pub(crate) fn take_drawable(&self, index: hal::SwapImageIndex) -> Result<metal::Drawable, ()> {
         let mut frame = self
             .frames[index as usize]
             .inner
             .lock();
-        assert!(!frame.available);
-        frame.available = true;
-        frame.drawable
-            .take()
-            .expect("Drawable has not been acquired!")
+        assert!(!frame.available && frame.linked);
+
+        match frame.drawable.take() {
```

Comment:
Could we use `ok_or` here?

Reply:
we don't want to set the `available` state unconditionally, so I don't think `ok_or` would help here